src/api-service/__app__/onefuzzlib/azure/nsg.py

```diff
@@ -0,0 +1,309 @@
+    resource_group = get_base_resource_group()
+    nsg = get_nsg(name)
+    if not nsg:
+        return Error(
+            code=ErrorCode.UNABLE_TO_FIND,
+            errors=["cannot update nsg rules. nsg %s not found" % name],
+        )
+
+    logging.info(
+        "setting allowed incoming connection sources for nsg: %s %s",
+        resource_group,
+        name,
+    )
+    security_rules = []
+    min_priority = 100
+    max_priority = 4096
+    if len(sources) > max_priority - min_priority:
+        return Error(
+            code=ErrorCode.INVALID_REQUEST,
+            errors=[
+                "too many rules provided %d. Max allowed: %d"
+                % ((len(sources)), max_priority - min_priority),
+            ],
+        )
+
+    priority = min_priority
+    for src in sources:
+        security_rules.append(
+            SecurityRule(
+                name="Allow" + str(priority),
+                protocol=SecurityRuleProtocol.TCP,
+                source_port_range="*",
+                destination_port_range="*",
+                source_address_prefix=src,
+                destination_address_prefix="*",
+                access=SecurityRuleAccess.ALLOW,
+                priority=priority,  # between 100 and 4096
+                direction="Inbound",
+            )
+        )
+        priority += 1
+
+    nsg.security_rules = security_rules
+    return update_nsg(nsg)
+
+
+def clear_all_rules(name: str) -> Union[None, Error]:
+    return set_allowed(name, [])
+
+
+def get_all_rules(name: str) -> Any:
+    nsg = get_nsg(name)
+    if not nsg:
+        return Error(
+            code=ErrorCode.UNABLE_TO_FIND,
+            errors=["cannot get nsg rules. nsg %s not found" % name],
+        )
+
+    return nsg.security_rules
+
+
+def associate_nic(name: str, nic: NetworkInterface) -> Union[None, Error]:
+    resource_group = get_base_resource_group()
+    nsg = get_nsg(name)
+    if not nsg:
+        return Error(
+            code=ErrorCode.UNABLE_TO_FIND,
+            errors=["cannot associate nic. nsg %s not found" % name],
+        )
+
+    if nsg.location != nic.location:
+        return Error(
+            code=ErrorCode.UNABLE_TO_UPDATE,
+            errors=[
+                "network interface and nsg have to be in the same region.",
+                "nsg %s %s, nic: %s %s"
+                % (nsg.name, nsg.location, nic.name, nic.location),
+            ],
+        )
+
+    if nic.network_security_group and nic.network_security_group.id == nsg.id:
+        return None
+
+    logging.info("associating nic %s with nsg: %s %s", nic.name, resource_group, name)
+
+    nic.network_security_group = nsg
+    network_client = get_network_client()
+    try:
+        network_client.network_interfaces.begin_create_or_update(
+            resource_group, nic.name, nic
+        )
+    except (ResourceNotFoundError, CloudError) as err:
+        if "The request failed due to conflict with a concurrent request" in str(err):
```

Comment:
Apply earlier comments about identifying concurrent request conflicts, and our success conditions.

Reply:
https://github.com/microsoft/onefuzz/issues/1349